swarm/src/connection.rs

```diff
@@ -286,24 +301,35 @@ where
                 Poll::Ready(ConnectionHandlerEvent::ReportRemoteProtocols(
                     ProtocolSupport::Added(protocols),
                 )) => {
-                    if let Some(added) =
-                        ProtocolsChange::add(remote_supported_protocols, &protocols)
-                    {
-                        handler.on_connection_event(ConnectionEvent::RemoteProtocolsChange(added));
-                        remote_supported_protocols.extend(protocols);
+                    let added = protocols
+                        .into_iter()
+                        .filter(|i| remote_supported_protocols.insert(i.clone()))
+                        .collect::<Vec<_>>();
+
+                    if !added.is_empty() {
+                        handler.on_connection_event(ConnectionEvent::RemoteProtocolsChange(
+                            ProtocolsChange::Added(ProtocolsAdded {
+                                protocols: added.iter(),
+                            }),
+                        ));
                     }
 
                     continue;
                 }
                 Poll::Ready(ConnectionHandlerEvent::ReportRemoteProtocols(
                     ProtocolSupport::Removed(protocols),
                 )) => {
-                    if let Some(removed) =
-                        ProtocolsChange::remove(remote_supported_protocols, &protocols)
-                    {
-                        handler
-                            .on_connection_event(ConnectionEvent::RemoteProtocolsChange(removed));
-                        remote_supported_protocols.retain(|p| !protocols.contains(p));
+                    let removed = protocols
+                        .into_iter()
+                        .filter_map(|i| remote_supported_protocols.take(&i))
```

Comment:
Why `take` and not just `remove` and `filter`?

Reply:
no particular reason, both solutions are valid, I picked this one